gles3_builders.py

```diff
@@ -189,399 +188,334 @@ def include_file_in_gles3_header(filename: str, header_data: GLES3HeaderStruct,
-                + str(defspec)
-                + """) {  _FU
-
-                const Transform2D &tr = p_transform;
-
-            GLfloat matrix[16]={ /* build a 16x16 matrix */
-                (GLfloat)tr.columns[0][0],
-                (GLfloat)tr.columns[0][1],
-                (GLfloat)0,
-                (GLfloat)0,
-                (GLfloat)tr.columns[1][0],
-                (GLfloat)tr.columns[1][1],
-                (GLfloat)0,
-                (GLfloat)0,
-                (GLfloat)0,
-                (GLfloat)0,
-                (GLfloat)1,
-                (GLfloat)0,
-                (GLfloat)tr.columns[2][0],
-                (GLfloat)tr.columns[2][1],
-                (GLfloat)0,
-                (GLfloat)1
-            };
-
-                glUniformMatrix4fv(version_get_uniform(p_uniform,p_version,p_variant,p_specialization),1,false,matrix);
-
-            }
-
-            """
+        if header_data.specialization_names:
+            specialization_names = ",\n\t\t".join(
+                f"{name.upper()} = {1 << index}" for index, name in enumerate(header_data.specialization_names)
             )
+            file.write(f"""\
+	enum Specializations {{
+		{specialization_names}
+	}};
 
-            fd.write(
-                """_FORCE_INLINE_ void version_set_uniform(Uniforms p_uniform, const Projection& p_matrix, RID p_version, ShaderVariant p_variant"""
-                + defvariant
-                + """,uint64_t p_specialization="""
-                + str(defspec)
-                + """) {  _FU
+""")
+        for index, specialization_value in enumerate(header_data.specialization_values):
+            if specialization_value.strip().upper() in ["TRUE", "1"]:
+                defspec |= 1 << index
 
-                GLfloat matrix[16];
+        file.write(f"""\
+	_FORCE_INLINE_ bool version_bind_shader(RID p_version, ShaderVariant p_variant{defvariant}, uint64_t p_specialization = {defspec}) {{
+		return _version_bind_shader(p_version, p_variant, p_specialization);
+	}}
 
-                for (int i = 0; i < 4; i++) {
-                    for (int j = 0; j < 4; j++) {
-                        matrix[i * 4 + j] = p_matrix.columns[i][j];
-                    }
-                }
+""")
 
-                glUniformMatrix4fv(version_get_uniform(p_uniform, p_version, p_variant, p_specialization), 1, false, matrix);
-        }"""
-            )
-
-            fd.write("\n\n#undef _FU\n\n\n")
-
-        fd.write("protected:\n\n")
-
-        fd.write("\tvirtual void _init() override {\n\n")
+        if header_data.uniforms:
+            file.write(f"""\
+	_FORCE_INLINE_ int version_get_uniform(Uniforms p_uniform, RID p_version, ShaderVariant p_variant{defvariant}, uint64_t p_specialization = {defspec}) {{
+		return _version_get_uniform(p_uniform, p_version, p_variant, p_specialization);
+	}}
+
+	/* clang-format off */
+#define TRY_GET_UNIFORM if (version_get_uniform(p_uniform, p_version, p_variant, p_specialization) < 0) return
```

Comment:
This seems like a change in behaviour (unless I'm missing something).
In this case I'd like to see an approval from the shader team as well.
I'm specifically sceptical if silently ignoring unknown uniforms is a good idea?

Reply:
Actually, no! It's the same define setup here, just given a clearer name
https://github.com/godotengine/godot/blob/2303ce843a362cf5497ca3336451de23793eb711/gles3_builders.py#L272-L274

Reply:
Oh wow, `_FU` was so short I just missed it! Nevermind then :smile: